azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,692 @@
+        '''Uses device info from file to recreate the hub state'''
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+
+        self.delete_aspects(replace, hub_aspects)
+
+        try:
+            with open(state_file, 'r', encoding='utf-8') as f:
+                hub_state = json.load(f)
+
+            self.upload_hub_from_dict(hub_state, hub_aspects)
+            logger.info("Uploaded state from '{}' to IoT Hub '{}'".format(state_file, self.hub_name))
+
+        except FileNotFoundError:
+            raise FileOperationError(f'File {state_file} does not exist.')
+
+    def migrate_state(
+        self,
+        orig_hub: Optional[str] = None,
+        orig_rg: Optional[str] = None,
+        orig_hub_login: Optional[str] = None,
+        replace: Optional[bool] = False,
+        hub_aspects: List[str] = None
+    ):
+        '''Migrates state from original hub to destination hub.'''
+        orig_hub_target = self.discovery.get_target(
+            resource_name=orig_hub,
+            resource_group_name=orig_rg,
+            login=orig_hub_login,
+            auth_type=self.auth_type
+        )
+
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+
+        # Command modifies hub_aspect - make copy so we can reuse for upload
+        hub_state = self.process_hub_to_dict(orig_hub_target, hub_aspects[:])
+        self.delete_aspects(replace, hub_aspects)
+        self.upload_hub_from_dict(hub_state, hub_aspects)
+
+        logger.info("Migrated state from IoT Hub '{}' to {}".format(orig_hub, self.hub_name))
+
+    def delete_aspects(self, replace, hub_aspects=[]):
+        """
+        Delete all necessary hub aspects if the hub exists.
+
+        Delete dataplane aspects only present in hub_aspects.
+        """
+        if self.target and replace:
+            if HubAspects.Configurations.value in hub_aspects:
+                self.delete_all_configs()
+            if HubAspects.Devices.value in hub_aspects:
+                self.delete_all_devices()
+            if HubAspects.Arm.value in hub_aspects:
+                cert_client = self.get_client().certificates
+                # serialize strips name and etag - use as_dict instead
+                certificates = cert_client.list_by_iot_hub(self.rg, self.hub_name).as_dict()
+
+                for cert in tqdm(certificates["value"], desc="Deleting certificates from destination hub", ascii=" #"):
+                    cert_client.delete(self.rg, self.hub_name, cert["name"], cert["etag"])
+
+    def process_hub_to_dict(self, target: dict, hub_aspects: list = []) -> dict:
+        '''Returns a dictionary containing the hub state'''
+        hub_state = {}
+
+        if HubAspects.Configurations.value in hub_aspects:
+            hub_aspects.remove(HubAspects.Configurations.value)
+            # Basic tier does not support list config
+            try:
+                all_configs = _iot_hub_configuration_list(target=self.target)
```

Comment:
This will throw a stacktrace when attempting to export from a non-existent hub, for example:

az iot hub state export --aspects configurations -f /file/location/state.json -n NonExistentHub -g MyRG

nets a stack trace with

![image](https://user-images.githubusercontent.com/28658112/197652759-23be28c5-a537-47a1-963a-8888b4e92066.png)

Lets please add some test coverage for these kind of failure modes

Reply:
fixed, will add tests for failure cases